executor/infoschema_reader.go

```diff
@@ -153,3 +162,164 @@ func dataForViews(ctx sessionctx.Context, schemas []*model.DBInfo) ([][]types.Da
+		e.retrieved = true
+		return nil, nil
+	}
+
+	rows, err := e.dataForDiskUsage(sctx)
+	if err != nil {
+		return nil, err
+	}
+	if len(e.outputCols) == len(e.table.Columns) {
+		return rows, nil
+	}
+	retRows := make([][]types.Datum, len(rows))
+	for i, fullRow := range rows {
+		row := make([]types.Datum, len(e.outputCols))
+		for j, col := range e.outputCols {
+			row[j] = fullRow[col.Offset]
+		}
+		retRows[i] = row
+	}
+	return retRows, nil
+}
+
+type initialTable struct {
+	db string
+	tb *model.TableInfo
+}
+
+func (e *diskUsageRetriever) initialize(sctx sessionctx.Context) error {
+	is := infoschema.GetInfoSchema(sctx)
+	dbs := is.AllSchemas()
+	sort.Sort(infoschema.SchemasSorter(dbs))
+	schemas := e.extractor.TableSchema
+	tables := e.extractor.TableName
+	//var initialTables []initialTable
+	if len(schemas) != 0 {
+		for _, schema := range dbs {
+			if schemas.Exist(schema.Name.L) {
+				if len(tables) != 0 {
+					for _, table := range schema.Tables {
+						if tables.Exist(table.Name.L) {
+							e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+						}
+					}
+				} else {
+					for _, table := range schema.Tables {
+						e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+					}
+				}
+			}
+
+		}
+		if len(e.initialTables) == 0 && tables == nil {
+			return errors.Errorf("schema or table not exist, please check the schema and table")
+		}
+	} else {
+		if len(tables) != 0 {
+			for _, schema := range dbs {
+				for _, table := range schema.Tables {
+					if tables.Exist(table.Name.L) {
+						e.initialTables = append(e.initialTables, &initialTable{db: schema.Name.L, tb: table})
+					}
+				}
+			}
+		} else {
+			for _, schema := range dbs {
+				for _, table := range schema.Tables {
+					e.initialTables = append(e.initialTables, &initialTable{schema.Name.O, table})
+				}
+			}
+		}
+	}
+	e.initialized = true
+	return nil
+}
+
+// pdRegionStats is the json response from PD.
+type pdRegionStats struct {
+	Count            int              `json:"count"`
+	EmptyCount       int              `json:"empty_count"`
+	StorageSize      int64            `json:"storage_size"`
+	StoreLeaderCount map[uint64]int   `json:"store_leader_count"`
+	StorePeerCount   map[uint64]int   `json:"store_peer_count"`
+	StoreLeaderSize  map[uint64]int64 `json:"store_leader_size"`
+	StorePeerSize    map[uint64]int64 `json:"store_peer_size"`
+}
+
+func (e *diskUsageRetriever) dataForDiskUsage(ctx sessionctx.Context) ([][]types.Datum, error) {
+	rows := make([][]types.Datum, 0, 1024)
+	tikvStore, ok := ctx.GetStore().(tikv.Storage)
+	if !ok {
+		return nil, errors.New("Information about TiKV region status can be gotten only when the storage is TiKV")
+	}
+	var pdAddrs []string
+	etcd, ok := tikvStore.(tikv.EtcdBackend)
+	if !ok {
+		return nil, errors.New("not implemented")
+	}
+	pdAddrs = etcd.EtcdAddrs()
+	if len(pdAddrs) < 0 {
+		return nil, errors.New("pd unavailable")
+	}
```

Comment:
This block seems to appear many times in many places, could you refactor it?

Reply:
em...,this can be refactor in another PR.